### PR TITLE
[NativeAOT] Reflection Invoke refactoring

### LINKED ISSUE #72548: Cannot box a ByRef-like (ref struct) when using a CreateDelegate method on method that accepts a ByRef-like parameter.
The following code correctly runs on CoreCLR and fails on NativeAOT:
[https://gist.github.com/Suchiman/b09ad481065e3e2b17d89e321121bacc](url)

NativeAOT fails with the following exception: `System.NotSupportedException: Cannot create boxed ByRef-like values.`

The issue occurs at this line: 
` typeof(Program).GetMethod("Method")!.CreateDelegate<ExampleDelegate>();`

### PR DESCRIPTION
**Pull request by jkotas**
- Refactored reflection Invoke in NativeAOT to be similar to how it is done in CoreCLR.
- All argument validation and coercion is done via static code now. This makes the reflection invoke stubs much smaller and the AOT compiler side a lot simpler.
- The invoke stub is minimal now and just takes the arguments for the methods to invoke as "Span of byrefs". The two notable differences with CoreCLR are:
   - The invoke stub takes the function pointer to call as an argument to allow sharing of the stubs between methods with the same signature. CoreCLR has the thunks non-sharable currently. We have discussed sharing them among methods with the same signature like it is done here.
   - The return value is returned as byref. CoreCLR thunk does boxing of the return value as part of the stub. Again, we have discussed to do it this way in CoreCLR too, we just did not have time to do it yet.

Fixes #72548

### CODE REVIEW

**Comment by jkotas** [comment]
Some perf numbers
  | NativeAOT - current | NativeAOT - this PR | CoreCLR - net6.0 | CoreCLR - net7.0
-- | -- | -- | -- | --
Invoke static method with no arguments | 22ns | 11ns | 34ns | 8.2ns
Invoke instance method with int and   string arguments | 67ns | 34ns | 96ns | 38ns


The extra fixed overhead that is observable in invoke static method with no arguments is due extra "Invoker" layer in native AOT. I am going to delete that layer in subsequent PRs.

**Comment by jkotas** [comment]
cc @steveharter @dotnet/area-system-reflection

**Comment by jkotas** [comment]
@MichalStrehovsky @dotnet/ilc-contrib This is ready for review.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+            {
+                ret = ref InvokeWithManyArguments( methodToCall, ref thisArg, ref ret,
+                    parameters, binderBundle, wrapInTargetInvocationException);
+            }
+            else
+            {
+                StackAllocedArguments argStorage = default;
+                StackAllocatedByRefs byrefStorage = default;
+
+                CheckArguments( ref argStorage._arg0!, (ByReference*)&byrefStorage, parameters, binderBundle);
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs:238]:
```suggestion
                CheckArguments(ref argStorage._arg0!, (ByReference*)&byrefStorage, parameters, binderBundle);
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+                }
+                finally
+                {
+                    if (_needsCopyBack)
+                        CopyBack( ref argStorage._arg0!, parameters);
+                }
+            }
+
+            return ((_returnTransform & (Transform.Nullable | Transform.Pointer | Transform.ByRef)) != 0) ?
+                ReturnTranform( ref ret, wrapInTargetInvocationException) : returnObject;
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs:257]:
```suggestion
                ReturnTranform(ref ret, wrapInTargetInvocationException) : returnObject;
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+
+            RuntimeImports.GCFrameRegistration regArgStorage = new(pStorage, (uint)argCount, areByRefs: false);
+            RuntimeImports.GCFrameRegistration regByRefStorage = new(pByRefStorage, (uint)argCount, areByRefs: true);
+
+            try
+            {
+                RuntimeImports.RhRegisterForGCReporting(&regArgStorage);
+                RuntimeImports.RhRegisterForGCReporting(&regByRefStorage);
+
+                CheckArguments( ref Unsafe.As<IntPtr, object>(ref *pStorage), pByRefStorage, parameters, binderBundle);
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs:281]:
```suggestion
                CheckArguments(ref Unsafe.As<IntPtr, object>(ref *pStorage), pByRefStorage, parameters, binderBundle);
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+                    {
+                        obj = RuntimeImports.RhBox(argumentInfo.Type, ref obj.GetRawData());
+                    }
+                }
+
+                parameters[i] = obj;
+            }
+        }
+
+        private unsafe object ReturnTranform( ref byte byref, bool wrapInTargetInvocationException)
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs:438]:
```suggestion
        private unsafe object ReturnTransform( ref byte byref, bool wrapInTargetInvocationException)
```

**Review by MichalStrehovsky — APPROVED** [review]
Very nice!

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I think this broke the API compat for AOT.

```
  LibraryImportGenerator -> ...\runtime\artifacts\bin\LibraryImportGenerator\Release\netstandard2.0\Microsoft.Interop.LibraryImportGenerator.dll
  System.Private.CoreLib -> ...\runtime\artifacts\bin\System.Private.CoreLib\ref\Release\net7.0\System.Private.CoreLib.dll
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : Compat issues with assembly System.Private
.CoreLib: [...\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.csproj]
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : TypesMustExist : Type 'System.Reflection.D 
ynamicInvokeInfo' does not exist in the reference but it does exist in the implementation. [...\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.cs 
proj]
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error :  [...\runtime\src\coreclr\nativeaot\System. 
Private.CoreLib\src\System.Private.CoreLib.csproj]
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : *** Invalid/Unused baseline differences ** 
* [...\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.csproj]
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : TypesMustExist : Type 'System.InvokeUtils' 
 does not exist in the reference but it does exist in the implementation. [...\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.csproj]
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : TypesMustExist : Type 'System.Reflection.D 
elegateDynamicInvokeInfo' does not exist in the reference but it does exist in the implementation. [...\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.Co 
reLib.csproj]
...\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22372.1\build\Microsoft.DotNet.ApiCompat.targets(112,5): error : MatchingRefApiCompat failed - The referen 
ce assembly doesn't match all the APIs in the implementation for '...\runtime\artifacts\bin\coreclr\windows.x64.Release\aotsdk\System.Private.CoreLib.dll'. To address either fix  
errors in the reference assembly (referenced as implementation in compat errors for this reverse compat check), add the issues to the baseline file '...\runtime\src\coreclr\nativ 
eaot\System.Private.CoreLib\src\MatchingRefApiCompatBaseline.txt' or disable this check by setting RunMatchingRefApiCompat=false in this project. [...\runtime\src\coreclr\nativea 
ot\System.Private.CoreLib\src\System.Private.CoreLib.csproj]
```

**Comment by jkotas** [comment]
@akoeplinger It looks like your https://github.com/dotnet/runtime/pull/72143 was not updated to account for my changes. Could you please fix it up?

**Comment by akoeplinger** [comment]
Yep, I did merge main into the branch but looks like it was a few minutes before your PR went it.

**Comment by akoeplinger** [comment]
Opened a PR with the fix: https://github.com/dotnet/runtime/pull/73248

**Pull request merged**